tests/test_complex_types.cpp

```diff
@@ -2,120 +2,75 @@
+  }
 
-TEST_MATH_OP_TYPE(add, +)
-TEST_MATH_OP_TYPE(sub, -)
-TEST_MATH_OP_TYPE(mul, *)
-TEST_MATH_OP_TYPE(div, /)
+TEST_MATH_OP_TYPE("Test complex addition types", "[add]", +)
+TEST_MATH_OP_TYPE("Test complex subtraction types", "[sub]", -)
+TEST_MATH_OP_TYPE("Test complex multiplication types", "[mul]", *)
+TEST_MATH_OP_TYPE("Test complex division types", "[div]", /)
 #undef TEST_MATH_FUNC_TYPE
 
-// Check operations return correct types
-void check_math_operator_types() {
-  test_valid_types<test_add_types>();
-  test_valid_types<test_sub_types>();
-  test_valid_types<test_mul_types>();
-  test_valid_types<test_div_types>();
-}
-
 // Define math function tests
-#define TEST_MATH_FUNC_TYPE(func)                                              \
-  template <typename T> struct test##_##func##_##types {                       \
-    bool operator()() {                                                        \
-      static_assert(std::is_same_v<complex<T>, decltype(func(complex<T>()))>); \
-      return true;                                                             \
-    }                                                                          \
-  };
-
-TEST_MATH_FUNC_TYPE(acos)
-TEST_MATH_FUNC_TYPE(asin)
-TEST_MATH_FUNC_TYPE(atan)
-TEST_MATH_FUNC_TYPE(acosh)
-TEST_MATH_FUNC_TYPE(asinh)
-TEST_MATH_FUNC_TYPE(atanh)
-TEST_MATH_FUNC_TYPE(conj)
-TEST_MATH_FUNC_TYPE(cos)
-TEST_MATH_FUNC_TYPE(cosh)
-TEST_MATH_FUNC_TYPE(exp)
-TEST_MATH_FUNC_TYPE(log)
-TEST_MATH_FUNC_TYPE(log10)
-TEST_MATH_FUNC_TYPE(proj)
-TEST_MATH_FUNC_TYPE(sin)
-TEST_MATH_FUNC_TYPE(sinh)
-TEST_MATH_FUNC_TYPE(sqrt)
-TEST_MATH_FUNC_TYPE(tan)
-TEST_MATH_FUNC_TYPE(tanh)
-#undef TEST_MATH_FUNC_TYPE
-
-template <typename T> struct test_abs_types {
-  bool operator()() {
-    static_assert(std::is_same_v<T, decltype(abs(complex<T>()))>);
-    return true;
+#define TEST_MATH_FUNC_TYPE(test_name, label, func)                            \
+  TEMPLATE_TEST_CASE(test_name, label, double, float, sycl::half) {            \
+    static_assert(std::is_same_v<complex<TestType>,                            \
+                                 decltype(func(complex<TestType>()))>);        \
   }
-};
 
-template <typename T> struct test_polar_types {
-  bool operator()() {
-    static_assert(std::is_same_v<complex<T>, decltype(polar(T()))>);
-    static_assert(std::is_same_v<complex<T>, decltype(polar(T(), T()))>);
-    return true;
-  }
-};
-
-template <typename T> struct test_pow_types {
-  bool operator()() {
-    static_assert(std::is_same_v<complex<T>, decltype(pow(complex<T>(), T()))>);
-    static_assert(
-        std::is_same_v<complex<T>, decltype(pow(complex<T>(), complex<T>()))>);
-    static_assert(std::is_same_v<complex<T>, decltype(pow(T(), complex<T>()))>);
-    return true;
-  }
-};
-
-// Check functions return correct types
-void check_math_function_types() {
+TEST_MATH_FUNC_TYPE("Test complex acos types", "[acos]", acos)
```

Comment:
Maybe a stupid question, but it seems strange to me that if we list all the `TEST_MATH_FUNC_TYPE` we need to add them manually in the CMAKELIST. 
Sorry, I'm sure I'm missing something obvious...

Reply:
So I don't quite understand. `TEST_MATH_FUNC_TYPE` is a macro that just allows the test to be repeated for all the different functions without me re-writing them. CMake needs a list of all the source files which then Catch2 parses for the tests Catch2 looks for a different macro something along the line of `TEST_CASE("Test is_gencomplex", "[gencomplex]")`.
CMake needs a list of the test source files so it knows which files to compile otherwise it is will compile none of the tests files.
Does this answer your question, could you clarify if it does not?

Reply:
Oh I see. Yes it clarifies; thanks! `TEST_MATH_FUNC_TYPE(acos)` is not calling or registering the tests inside `acos_complex.cpp`. 

TEST_MATH_FUNC_TYPE are just another type of tests. 

Thanks!